File: app/src/main/java/com/synapse/social/studioasinc/util/AttachmentUtils.java
/**
 * CONFIDENTIAL AND PROPRIETARY
 * 
 * This source code is the sole property of StudioAs Inc. Synapse. (Ashik).
 * Any reproduction, modification, distribution, or exploitation in any form
 * without explicit written permission from the owner is strictly prohibited.
 * 
 * Copyright (c) 2025 StudioAs Inc. Synapse. (Ashik)
 * All rights reserved.
 */

package com.synapse.social.studioasinc.util;

import com.synapse.social.studioasinc.model.Attachment;

import java.util.ArrayList;
import java.util.HashMap;

/**
 * Utility class for converting between HashMap-based attachments and type-safe Attachment objects.
 * Provides migration path from legacy HashMap format to new Parcelable format.
 */
public class AttachmentUtils {

    /**
     * Converts a HashMap representation to an Attachment object.
     * 
     * @param attachmentMap The HashMap containing attachment data
     * @return Attachment object or null if conversion fails
     */
    public static Attachment fromHashMap(HashMap<String, Object> attachmentMap) {
        if (attachmentMap == null) {
            return null;
        }
        
        try {
            Attachment attachment = new Attachment();
            
            // Extract publicId
            Object publicIdObj = attachmentMap.get("publicId");
            if (publicIdObj != null) {
                attachment.setPublicId(String.valueOf(publicIdObj));
            }
            
            // Extract URL
            Object urlObj = attachmentMap.get("url");
            if (urlObj != null) {
                attachment.setUrl(String.valueOf(urlObj));
            }
            
            // Extract dimensions
            Object widthObj = attachmentMap.get("width");
            if (widthObj instanceof Number) {
                attachment.setWidth(((Number) widthObj).intValue());
            }
            
            Object heightObj = attachmentMap.get("height");
            if (heightObj instanceof Number) {
                attachment.setHeight(((Number) heightObj).intValue());
            }
            
            // Extract type
            Object typeObj = attachmentMap.get("type");
            if (typeObj != null) {
                attachment.setType(String.valueOf(typeObj));
            } else {
                // Infer type from publicId if not explicitly set
                String publicId = attachment.getPublicId();
                if (publicId != null && publicId.contains("|video")) {
                    attachment.setType("video");
                } else {
                    attachment.setType("image");
                }
            }
            
            // Extract mimeType
            Object mimeTypeObj = attachmentMap.get("mimeType");
            if (mimeTypeObj != null) {
                attachment.setMimeType(String.valueOf(mimeTypeObj));
            }
            
            // Extract size
            Object sizeObj = attachmentMap.get("size");
            if (sizeObj instanceof Number) {
                attachment.setSize(((Number) sizeObj).longValue());
            }
            
            return attachment;
            
        } catch (Exception e) {
            // Log error but don't crash
            android.util.Log.e("AttachmentUtils", "Error converting HashMap to Attachment: " + e.getMessage());
            return null;
        }
    }
    
    /**
     * Converts an Attachment object to a HashMap representation.
     * Used for backward compatibility with existing code.
     * 
     * @param attachment The Attachment object
     * @return HashMap representation
     */
    public static HashMap<String, Object> toHashMap(Attachment attachment) {
        if (attachment == null) {
            return null;
        }
        
        HashMap<String, Object> map = new HashMap<>();
        
        if (attachment.getPublicId() != null) {
            map.put("publicId", attachment.getPublicId());
        }
        
        if (attachment.getUrl() != null) {
            map.put("url", attachment.getUrl());
        }
        
        map.put("width", attachment.getWidth());
        map.put("height", attachment.getHeight());
        
        if (attachment.getType() != null) {
            map.put("type", attachment.getType());
        }
        
        if (attachment.getMimeType() != null) {
            map.put("mimeType", attachment.getMimeType());
        }
        
        map.put("size", attachment.getSize());
        
        return map;
    }
    
    /**
     * Converts a list of HashMap attachments to a list of Attachment objects.
     * 
     * @param attachmentMaps List of HashMap representations
     * @return List of Attachment objects
     */
    public static ArrayList<Attachment> fromHashMapList(ArrayList<HashMap<String, Object>> attachmentMaps) {
        if (attachmentMaps == null) {
            return new ArrayList<>();
        }
        
        ArrayList<Attachment> attachments = new ArrayList<>();
        
        for (HashMap<String, Object> map : attachmentMaps) {
            Attachment attachment = fromHashMap(map);
            if (attachment != null) {
                attachments.add(attachment);
            }
        }
        
        return attachments;
    }
    
    /**
     * Converts a list of Attachment objects to a list of HashMap representations.
     * 
     * @param attachments List of Attachment objects
     * @return List of HashMap representations
     */
    public static ArrayList<HashMap<String, Object>> toHashMapList(ArrayList<Attachment> attachments) {
        if (attachments == null) {
            return new ArrayList<>();
        }
        
        ArrayList<HashMap<String, Object>> maps = new ArrayList<>();
        
        for (Attachment attachment : attachments) {
            HashMap<String, Object> map = toHashMap(attachment);
            if (map != null) {
                maps.add(map);
            }
        }
        
        return maps;
    }
}

File: app/src/main/java/com/synapse/social/studioasinc/util/AttachmentUtils.kt
/**
 * CONFIDENTIAL AND PROPRIETARY
 *
 * This source code is the sole property of StudioAs Inc. Synapse. (Ashik).
 * Any reproduction, modification, distribution, or exploitation in any form
 * without explicit written permission from the owner is strictly prohibited.
 *
 * Copyright (c) 2025 StudioAs Inc. Synapse. (Ashik)
 * All rights reserved.
 */

package com.synapse.social.studioasinc.util

import com.synapse.social.studioasinc.model.Attachment
import java.util.ArrayList
import java.util.HashMap

/**
 * Utility class for converting between HashMap-based attachments and type-safe Attachment objects.
 * Provides migration path from legacy HashMap format to new Parcelable format.
 */
object AttachmentUtils {

    /**
     * Converts a HashMap representation to an Attachment object.
     *
     * @param attachmentMap The HashMap containing attachment data
     * @return Attachment object or null if conversion fails
     */
    fun fromHashMap(attachmentMap: HashMap<String, Any?>?): Attachment? {
        if (attachmentMap == null) {
            return null
        }

        try {
            val attachment = Attachment()

            // Extract publicId
            val publicIdObj = attachmentMap["publicId"]
            if (publicIdObj != null) {
                attachment.publicId = publicIdObj.toString()
            }

            // Extract URL
            val urlObj = attachmentMap["url"]
            if (urlObj != null) {
                attachment.url = urlObj.toString()
            }

            // Extract dimensions
            val widthObj = attachmentMap["width"]
            if (widthObj is Number) {
                attachment.width = widthObj.toInt()
            }

            val heightObj = attachmentMap["height"]
            if (heightObj is Number) {
                attachment.height = heightObj.toInt()
            }

            // Extract type
            val typeObj = attachmentMap["type"]
            if (typeObj != null) {
                attachment.type = typeObj.toString()
            } else {
                // Infer type from publicId if not explicitly set
                val publicId = attachment.publicId
                if (publicId != null && publicId.contains("|video")) {
                    attachment.type = "video"
                } else {
                    attachment.type = "image"
                }
            }

            // Extract mimeType
            val mimeTypeObj = attachmentMap["mimeType"]
            if (mimeTypeObj != null) {
                attachment.mimeType = mimeTypeObj.toString()
            }

            // Extract size
            val sizeObj = attachmentMap["size"]
            if (sizeObj is Number) {
                attachment.size = sizeObj.toLong()
            }

            return attachment

        } catch (e: Exception) {
            // Log error but don't crash
            android.util.Log.e("AttachmentUtils", "Error converting HashMap to Attachment: " + e.message)
            return null
        }
    }

    /**
     * Converts an Attachment object to a HashMap representation.
     * Used for backward compatibility with existing code.
     *
     * @param attachment The Attachment object
     * @return HashMap representation
     */
    fun toHashMap(attachment: Attachment?): HashMap<String, Any>? {
        if (attachment == null) {
            return null
        }

        val map = HashMap<String, Any>()

        attachment.publicId?.let { map["publicId"] = it }
        attachment.url?.let { map["url"] = it }

        map["width"] = attachment.width
        map["height"] = attachment.height

        attachment.type?.let { map["type"] = it }
        attachment.mimeType?.let { map["mimeType"] = it }

        map["size"] = attachment.size

        return map
    }

    /**
     * Converts a list of HashMap attachments to a list of Attachment objects.
     *
     * @param attachmentMaps List of HashMap representations
     * @return List of Attachment objects
     */
    fun fromHashMapList(attachmentMaps: ArrayList<HashMap<String, Any?>>?): ArrayList<Attachment> {
        if (attachmentMaps == null) {
            return ArrayList()
        }

        val attachments = ArrayList<Attachment>()

        for (map in attachmentMaps) {
            val attachment = fromHashMap(map)
            if (attachment != null) {
                attachments.add(attachment)
            }
        }

        return attachments
    }

    /**
     * Converts a list of Attachment objects to a list of HashMap representations.
     *
     * @param attachments List of Attachment objects
     * @return List of HashMap representations
     */
    fun toHashMapList(attachments: ArrayList<Attachment>?): ArrayList<HashMap<String, Any>> {
        if (attachments == null) {
            return ArrayList()
        }

        val maps = ArrayList<HashMap<String, Any>>()

        for (attachment in attachments) {
            val map = toHashMap(attachment)
            if (map != null) {
                maps.add(map)
            }
        }

        return maps
    }
}
